### ISSUE TITLE
Possible wrong timer behavoir in run_wasm.js

### ISSUE THREAD

**Issue by yurydelendik** [dart-lang/sdk#59666]
TL;DR I suspect a defect in the implementation of the addTimer code:

https://github.com/dart-lang/sdk/blob/f53c758b11bd17b37f5b5643d20c6a719812476e/pkg/dart2wasm/bin/run_wasm.js#L145

It is connected with `isNextTimerDue` logic: when `ms == 0` and `isNextTimerDue() == true`, the Task.run may create two separate entries in the timerHeap (in comparison with `isNextTimerDue() == false`, they will be just added to `zeroTimerQueue`) when engine runs slow.

The flute engine produces timers with the following cadence: 16 0 0 16 0 0 ..  in the [scheduleFrame](https://github.com/dart-lang/flute/blob/c44fba23c57bfc55031a823424112f77690aa00a/engine/lib/src/platform_dispatcher.dart#L567). The defect above causing dart benchmark to fail "Null check operator used on a null value" because the draw frame can be executed out of sequence with corresponding begin frame.

I'm temporary patching it with:

```js
    if (ms == 0){
      if (!isNextTimerDue()) {
        zeroTimerQueue.push(f);
      } else {
        timerHeap[0].push(f); // keeps Timer.run together
      } 
    } else {
```

**Comment by osa1** [comment]
Thanks for reporting. I don't understand how these callbacks can execute out of order given that earlier callbacks schedule the later ones:

https://github.com/dart-lang/flute/blob/c44fba23c57bfc55031a823424112f77690aa00a/engine/lib/src/platform_dispatcher.dart#L567-L593


```dart
void scheduleFrame() {
  _frameTimer ??= Timer(
    const Duration(milliseconds: 16),
    () {
      if (_continue) {
        _frameTimer = null;
      }
      final int microseconds = _frameTime.inMicroseconds;
      _frameTime += const Duration(milliseconds: 16);
      Timer.run(() {
        final Stopwatch beginSw = Stopwatch()..start();
        _beginFrame(microseconds);
        beginSw.stop();
        Timer.run(() {
          final Stopwatch drawSw = Stopwatch()..start();
          _drawFrame();
          drawSw.stop();
          _continue = frameCallback(
              _frameCount,
              beginSw.elapsedMicroseconds / 1000,
              drawSw.elapsedMicroseconds / 1000);
          _frameCount += 1;
        });
      });
    },
  );
}
```

Here `_drawFrame` is scheduled by the callback that calls `_beginFrame`, after calling `_beginFrame`.

Since there will be one callback scheduled by `scheduleFrame` at any point in time, do you mean callbacks of different `scheduleFrame`s interacting in a weird way maybe?

**Comment by yurydelendik** [comment]
I observe the failure in the benchmark only when I run it under valgrind. It is reproducible with debug builds of the JS engine.


> do you mean callbacks of different scheduleFrames interacting in a weird way maybe?

perhaps _beginFrame,_beginFrame,_drawFrame,_drawFrame

The `timerHeap` becomes something like [...30,...47,...31] in terms of timecodes,

Sorry, I wish I could be more helpful, I'm mostly working with wasm and js artifacts.

**Comment by lrhn** [comment]
That heap looks fine. It's a heap priority queue, so it implicitly represents subtrees through the indices. The child nodes of the node at position *i* are at positions *i*&times;2+1 and *i*&times;2+2. Here it makes `30` the root, with children `31` and `47`, both of which are larger numbers, so the heap is well-ordered.
When the root node is removed, the `31`-node will become the new root.

The way the timer queue implementation works is that it has a "due timers" list that it takes events from, and a heap priority queue of *later* timers, each a list containing the time they are due and a number of timers due at that time.
The heap queue root is always the timers with the next due time.

When the due timers list becomes empty, and a the next timer time in the heap is due, it's popped out of the heap, which is then normalized to put the next smallest time at the root.
The due time is removed from the front of the list and that list then becomes the new due-timer list, and event running can continue.

This means that all timers in the due-timer list are scheduled for *before* the current time. No new timer can be scheduled before them, because no new timer can be scheduled *before* the current time, it's no earlier than zero ms after the current `now`.

So when a new timer is added, it's added to the delayed timers priority heap, which puts it into the position it needs to be relative to all other existing scheduled timers. If any of those scheduled timers are also already due (but originally scheduled later than the ones in the due-timer list), the new event won't become the root of the priority heap.

The text `f (ms == 0 && !isNextTimerDue()) { ` is an optimization, it's not needed.
It checks whether the timer is a *zero-delay* timer, which means it will be scheduled for "`now`" and therefore already be due *and* that there is no due timer in the delayed timer heap, which means that it will *definitely* run before any of those timers, and definitely before any timers scheduled in the future.
In that case, it's added to the current due-timer list, to be run after all of those have been run.
Just an optimization, which shouldn't change when the event will run, it just avoids allocating a new list for and adding it to the heap (where it would become the root entry).

Without the `!isNextTimerDue()` check, adding the event to the due-timer list would make it run before the already due timers in the heap, which would be wrong.

The fix of
```dart
if (ms == 0){
      if (!isNextTimerDue()) {
        zeroTimerQueue.push(f);
      } else {
        timerHeap[0].push(f); // keeps Timer.run together
      } 
    } else {
```
is wrong too. It puts the zero-timer with the *first* due timers in the heap, but if there is more than one due timer-time in the heap, that will run the new timer *before* the second group of *already due* timers, which is incorrect.

**Comment by mkustermann** [comment]
Have sent out https://github.com/dart-lang/flute/pull/24 which I think is sufficient to fix this issue.

**Comment by mkustermann** [comment]
The flute benchmark was fixed some time ago. We've rolled that fix into our CI & benchmarking infrastructure some time ago.

Today I've also rolled it into https://github.com/mkustermann/wasm_gc_benchmarks, so this should close this bug.
